### PR TITLE
backup: Support --json flag by streaming JSON to report progress

### LINKED ISSUE #1937: Stream JSON output during long-running operations
Output of `restic version`
--------------------------

```
restic 0.9.2 (v0.9.2-0-gbeb208e1) compiled with go1.10.2 on darwin/amd64
```

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

When executing long-running commands such as `restic backup` or `restic restore`, I'd like to track the status programmatically (non-interactively) by having JSON stream through stdout. For example, during backup, a JSON object like this would be helpful (sans formatting, which I've added for readability):

```json
{
	"elapsed_time": "14:03",
	"estimated_time": "1:53:19",
	"percent_done": 0.6786,
	"total_files": 6223,
	"files_done": 1092,
	"total_bytes": 123456789,
	"bytes_done": 123456,
	"error_count": 0,
	"current_files": ["/my/file.txt", "/other/file.tar.gz"]
}
```

I figure an object like this could be written as often as the standard terminal UI is updated when in interactive mode.

Then I think something similar for restore, prune, check, and rebuild-index (in that order) would be useful -- but for now what matters most is backup.

I'm willing to submit the PR to do it, but I have to admit I looked through the [cmd_backup.go file](https://github.com/restic/restic/blob/beb208e1597a4b7d18705c12f045413056eafcdb/cmd/restic/cmd_backup.go#L362) and am a little lost at where the output is actually happening... there's a lot of indirection. Do you have any suggestions on how to go about this?

(This request is similar, but unrelated, to #1905)

Did restic help you or made you happy in any way?
-------------------------------------------------

Of course it has -- it's close to being ready for the rest of my family to use. :) Thanks for all the work you put into restic! I hope you find it as enjoyable to make as I do to use it.

### PR DESCRIPTION
**Pull request by mholt**
What is the purpose of this change? What does it change?
--------------------------------------------------------

`restic backup` now supports `--json` so that it can stream JSON output to update the progress of the backup, making it better for non-interactive, but supervised, use.

Currently, the basic updates are sent as JSON and seem to be working, but I am still transitioning the copy of ui.Backup to the new jsonstatus.Backup type. I will iterate on this.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Closes #1937

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by mholt** [comment]
I just realized my initial commit here writes out one JSON object per line, rather than streaming out an array of objects, as this example is designed to decode: https://godoc.org/encoding/json#example-Decoder-Decode-Stream

Will look into remedying that tomorrow... gotta decide if it's easier to decode a stream of objects like this or a single array that starts with `[` and streams out objects separated by commas, ending with `]`.

**Comment by mholt** [comment]
Alrighty -- I've finished implementing `restic --json backup` so that all the output is JSON-formatted. It also respects verbosity levels. Feel free to give it a spin!

Again, I don't feel this is the cleanest solution but I don't know a better way right now. Happy to have your thoughts or modifications if you have any ideas. :)

For the purposes of a convenient review, I've left in (comments) the original, human-formatted outputs so that you can compare what was being printed to the terminal and what is being populated in the JSON fields. They should be the same, sans fancy formatting.

Let me know what you think!

File: cmd/restic/global.go
Comment (mholt) [inline cmd/restic/global.go]: Just realized I made a similar, but slightly different variant of this change over at #1962: https://github.com/restic/restic/pull/1962/files#diff-952c97fd56de92e641e61b3b3aa881dbR363

**Comment by mholt** [comment]
Would anyone else like to try this out? I think it's pretty cool. I just don't love the code that does it. It'd be great to get another set of eyes on the code to get ready to merge this.

**Comment by Mebus** [comment]
Hi Matt,

I just tested it. This looks great! 

Can you please also include the ID of the created snapshot in the summary? This would be useful for a GUI.

```
→ restic -r /tmp/thebackup6 --json backup changelog
found 7 old cache directories in /home/mebus/.cache/restic, pass --cleanup-cache to remove them
{"message_type":"status","percent_done":0,"total_files":1,"total_bytes":364}
{"message_type":"status","percent_done":0.01380618243883937,"total_files":83,"files_done":1,"total_bytes":26365,"bytes_done":364,"current_files":["/changelog/0.6.0_2017-05-29/issue-953","/changelog/0.6.0_2017-05-29/issue-965"]}
{"message_type":"status","percent_done":0.14339015151515153,"total_files":140,"files_done":22,"total_bytes":52800,"bytes_done":7571,"current_files":["/changelog/0.7.2_2017-09-13/issue-1132","/changelog/0.7.2_2017-09-13/issue-1167"]}
{"message_type":"status","percent_done":0.34890151515151513,"total_files":140,"files_done":59,"total_bytes":52800,"bytes_done":18422,"current_files":["/changelog/0.8.1_2017-12-27/issue-1457"]}
{"message_type":"status","percent_done":1,"total_files":140,"files_done":140,"total_bytes":52800,"bytes_done":52800}
{"message_type":"summary","files_new":140,"files_changed":0,"files_unmodified":0,"dirs_new":0,"dirs_changed":0,"dirs_unmodified":0,"data_blobs":140,"tree_blobs":1,"data_added":53295,"total_files_processed":140,"total_bytes_processed":52800,"total_duration":0.7903416}
```
Thanks

Mebus

**Comment by mholt** [comment]
Sure thing @Mebus! Thanks for testing it. I'll revisit this PR shortly.

**Comment by mholt** [comment]
This PR is ready for approval and merge

**Comment by mholt** [comment]
@fd0 I know you're busy, but can you just click the Merge button? 😅 This is ready to go. :+1:

**Review by fd0 — APPROVED** [review]
LGTM!

**Comment by fd0** [comment]
I've rebased and resolved the merge conflict (and removed the merge of the master branch), and I'll merge this as soon as the CI tests complete :)

**Pull request merged**
